app-main에 스프링 시큐리티 + JWT를 활용한 로그인 테스트 코드 및 구현 코드 작성

kmmin78 opened the pull request:
주요 내용

1. 시큐리티 인증 구현체 코드 작성 및 빈 설정
2. jwt (rs256) 생성 유틸클래스 작성
3. 시큐리티 로그인의 end-to-end 테스트 코드 작성

Closes #56

Commit:
주요 내용

로그인 기능을 도메인 domain UseCase 인터페이스, 테스트 코드로 작성

외부 프레임워크, 인프라 의존성이 많은 app-main 모듈에서 해당 인터페이스를 의존한다.

Closes #50

Commit:
스프링 시큐리티 및 JWT 사용을 위한 라이브러리 의존성 추가 및 기반 설정 작업

주요 내용

1. 관련 라이브러리 의존성 추가 (spring security, spring boot test, jjwt, restAssured)
2. WebSecurity 설정 , app-core의 도메인 빈 설정.
  -> UserRepository를 빈 설정하기 위해 구현체 클래스 및 팩토리 메서드 생성

Closes #52

Commit:
app-main에 스프링 시큐리티 + JWT를 활용한 로그인 테스트 코드 및 구현 코드 작성

주요 내용

1. 시큐리티 인증 구현체 코드 작성 및 빈 설정
2. jwt (rs256) 생성 유틸클래스 작성
3. 시큐리티 로그인의 end-to-end 테스트 코드 작성

Closes #56

FrancescoJo commented:
이거 PR 이 큰 이유를 보니 #57 이 #53 의 기능을 포함하고 있어서 그렇네요. 이런 경우에는 PR 을 이렇게 만들어 주시는게 좋아요.

```
#53 <- #57
develop <- #53
```

FrancescoJo commented on app-main/src/main/kotlin/kr/flab/wiki/app/components/authentication/LoginUserService.kt:
```diff
@@ -0,0 +1,30 @@
+        val authentication: Authentication = authenticationManager.authenticate(
+            UsernamePasswordAuthenticationToken(
+                loginRequest.email,
+                loginRequest.password
```
서버의 메모리에 `password` 가 plaintext 형태로 남아있게 될 텐데, 이런 경우에는 어떻게 보안성을 높일 수 있을까요? 한번 생각해 보시는게 좋겠습니다.

kmmin78 replied:
저번에 말씀해주신 대로 password가 메모리에 남아있는 시간을 최대한 없애는 방향으로 가보겠습니다.
우선, 인증이 끝난 후의 UsernamePasswordAuthenticationToken에는 password 정보를 담지않는 것은 당연하고요.

FrancescoJo commented on app-main/src/main/kotlin/kr/flab/wiki/app/components/authentication/LoginUserService.kt:
```diff
@@ -0,0 +1,30 @@
+            return null
+        }
+        SecurityContextHolder.getContext().authentication = authentication
+        return LoginResponse(email = authentication.name, token = JwtUtils.generateJws(authentication))
```
`Service` 의 역할이 뭘까요?

응답을 만들어 주는건 `Service` 의 역할일까요, `Controller` 의 역할일까요?

kmmin78 replied:
응답은 Controller의 역할이 맞는데, 그럼 이 경우에는 Service에서 따로 DTO를 만들어 반환하는게 맞는걸까요?
LoginResponse와 DTO의 역할의 다른건가요...?

FrancescoJo commented on app-main/src/main/kotlin/kr/flab/wiki/app/utils/JwtUtils.kt:
```diff
@@ -0,0 +1,46 @@
+         * 해당 키페어는 서버 시작할 때마다 재생성됩니다.
+         * 추후에는 미리 생성한 후 설정파일(ex : application-*.yml)에서 읽어드릴 예정입니다.
+         */
+        private val KEY_PAIR: KeyPair = Keys.keyPairFor(SignatureAlgorithm.RS256)
```
여기서 질문:

키 생성시 실수로 보안성이 낮은 알고리즘을 이용한 바람에, 생성한 key 가 유출되거나, brute force 로 뚫려서 key 를 교체해야 하는 상황이 발생할 경우 어떻게 해결하는게 좋을까요?

재 배포 또는 서버 재시작 없이 key 를 교체할 방법이 있을까요?

kmmin78 replied:
재배포 혹은 서버 재시작 없이 key를 교체하는건 주석에 명시한대로 .yml파일 혹은 .properties 파일을 읽는 식으로 진행하면  따로 문제 없을 것 같은데, 다른 방법이 있을까요?

FrancescoJo commented on app-main/src/main/kotlin/kr/flab/wiki/app/utils/JwtUtils.kt:
```diff
@@ -0,0 +1,46 @@
+
+        fun generateJws(authentication: Authentication): String {
+            val email = authentication.name
+            val claims: Map<String, String> = mapOf("email" to email, "roles" to "")
```
jwt claims 은 외부에 노출될까요? 이 정보는 외부에 노출해도 괜찮은 정보일까요?

kmmin78 replied:
네 멘토링 시간에도 말씀드렸다시피, claims는 노출되는 정보(base64)이며, 노출해도 되는 정보만 담아야 합니다.

https://velog.io/@ypo09/JWTJSON-Web-Token-%EB%9E%80-%EB%AC%B4%EC%97%87%EC%9D%B8%EA%B0%80
위 링크를 참조하면 claim은
크게 registered claim, public claim, private claim의 3가지로 구분하는데 이건 분류상의 얘기일 뿐 claim 자체는 누구나 열어볼 수 있습니다.

FrancescoJo commented on app-main/src/test/kotlin/kr/flab/wiki/app/testcase/login/LoginWithSpringSecurityAndJwtTest.kt:
```diff
@@ -0,0 +1,169 @@
+            /**
+             * repository 에서 null 을 반환하도록 mocking 했는데, 정상적인 email, token값을 반환하는 현상 발생.
+             * 원인을 찾고 있음.
+             */
```
왜냐하면, `when` 으로 생성한 `mock` 이 실제 서비스 instance 에 주입되지 않아서 그렇습니다.

FrancescoJo replied:
이런 의도라면 MockMvcTest 를 써야 합니다. 하지만 이런 경우는 저라면 Spring restdocs 를 그대로 쓰되 @TestApplicationContext 에서만 동작하는 MockUserRepository 를 선언하고 이놈을 bean 으로 생성하도록 스프링 환경 설정을 바꿀 것 같아요.

kmmin78 replied:
이것도 테스트환경에서만 사용할 수 있는 MockUserRepository를 만드는 쪽으로 가겠습니다.

FrancescoJo commented on app-main/src/test/kotlin/kr/flab/wiki/app/testcase/login/LoginWithSpringSecurityAndJwtTest.kt:
```diff
@@ -0,0 +1,169 @@
+                fun `토큰을 반환한다`() {
+
+                    given()
+                        .accept(MediaType.APPLICATION_JSON_VALUE)
```
아래 method chain call 이 반복되는걸 짧게 줄일 방법이 있을까요?

kmmin78 replied:
방법을 찾아보겠습니다. ㅎㅎ

Commit:
app-main에 스프링 시큐리티 + JWT를 활용한 로그인 테스트 코드 및 구현 코드 작성
및 추가 수정

주요 내용

1. 테스트 코드 method chain 간소화 (with rest-assured:kotlin-extensions)
2. email, password 지역변수 제거 (보안)
3. 테스트 코드 정상동작 확인 (@MockBean 활용)

Closes #56
